languages/julia/exercises/concept/fibonacci-iterator/.docs/introduction.md

```diff
@@ -0,0 +1,113 @@
+## iterator-protocol
+
+<!-- TODO: Motivate why iterators are useful -->
+
+To implement the informal iteration interface, you need to define two methods `iterate(iter::MyIter)` and `iterate(iter::MyIter, state)`.
+Both methods must return a tuple of an item and the state of the iterator.
+The first method will return the first item and state, while the second method will return the next item and state.
+The iteration ends if `iterate` returns `nothing`.
+
+!!! note
+    Julia currently doesn't have a way to formally define interfaces.
+    To implement an interface, you need to look up which methods need to be defined in the Manual, or in the documentation of the package that "defines" the interface.
+    Alternatively, you can work your way through `MethodErrors` to find out which methods you need to implement:
+    <!-- TODO check if code blocks work within info boxes -->
+    ```julia
+    julia> struct MyIter end;
+
+    julia> for i in MyIter()
+            @show i
+        end
+    ERROR: MethodError: no method matching iterate(::MyIter)
+    ...
+    ```
+
+!!! note
+    Julia iterators track the state decoupled from the iterator object.
+    The iterator object itself is generally not mutated.
```

Comment:
```suggestion
!!! note
    The iterator object itself is usually not mutated by iteration
```

I think this is a little confusing. Maybe better to say less and be more correct.

Reply:
I'm happy to change it to your suggestion but could you explain why it's more correct?

I based that note on the [manual](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-iteration):

> Instead of mutating objects as they are iterated over, Julia iterators may keep track of the iteration state externally from the object.

Reply:
Yea, that sounds better :)